Default directory may start with tilde

tdwilson opened the pull request:
`arduino-cli--compile` works when `default-directory` starts with a tilde (as it can do on Linux Mint 22.1).
`arduino-cli--message` doesn't `shell-quote-argument` a path headed for `default-directory` (rather than a shell command line), and removes trailing blank lines from the output displayed in the minibuffer.

tdwilson committed:
Allow for default-directory to start with ~

This commit contains some changes I made that enable arduino-cli-mode
to work on Linux Mint 22.1 when the sketch's default directory name
starts with ~.  The README.md says that arduino-cli-mode was tested on
MacOS; unfortunately I don't have a MacOS system I can test these
changes against.

A buffer's default-directory may start with ~, as documented in the
output of (describe-variable 'default-directory), which starts with
the following lines [edited]:
----
default-directory is a variable defined in ‘C source code’.

Its value is
"~/.arduino15/RemoteSketchbook/ArduinoCloud/[...]/NM_RSSI_list/"
Local in buffer NM_RSSI_list.ino; global value is nil

Name of default directory of current buffer.
It should be an absolute directory name; on GNU and Unix systems,
these names start with "/" or "~" and end with "/".
----

A default-directory that starts with ~ causes arduino-cli-compile to
fail as shown in the contents of the *arduino-cli-compilation* buffer
[edited]:
----
-*- mode: arduino-cli-compilation; default-directory: "~/.arduino15/RemoteSketchbook/ArduinoCloud/[...]/NM_RSSI_list/" -*-
arduino-cli-compilation started at Wed Apr  9 20:07:31

arduino-cli compile --fqbn esp8266:esp8266:nodemcuv2 \~/.arduino15/RemoteSketchbook/ArduinoCloud/[...]/NM_RSSI_list/ -t --warnings all
Can't open sketch: no such file or directory: /home/tdw/.arduino15/RemoteSketchbook/ArduinoCloud/[...]/NM_RSSI_list/~/.arduino15/RemoteSketchbook/ArduinoCloud/[...]/NM_RSSI_list

arduino-cli-compilation exited abnormally with code 1 at Wed Apr  9 20:07:32
----

The reason arduino-cli-compilation fails is that a leading ~ in
default-directory is quoted by shell-quote-argument (as seen above)
giving a non-existent directory name.  The fix is to remove any
leading ~ using expand-file-name.

arduino-cli-compilation now behaves as expected:
----
-*- mode: arduino-cli-compilation; default-directory: "~/.arduino15/RemoteSketchbook/ArduinoCloud/[...]/NM_RSSI_list/" -*-
arduino-cli-compilation started at Wed Apr  9 20:37:05

arduino-cli compile --fqbn esp8266:esp8266:nodemcuv2 /home/tdw/.arduino15/RemoteSketchbook/ArduinoCloud/[...]/NM_RSSI_list/ -t --warnings all
[...]
----

A default-directory that starts with tilde also causes
arduino-cli--message to fail, eg as seen with arduino-cli-board-list,
which fails with the message:
----
shell-command: Setting current directory: No such file or directory, \~/.arduino15/RemoteSketchbook/ArduinoCloud/[...]/NM_RSSI_list/
----

The fix is similar: expand-file-name the default-directory (or
provided directory) before using it.  I have also deleted the
shell-quote-argument as the default-directory isn't passed to a shell.
If you leave the shell-quote-argument, then arduino-cli-board-list
with a sketch called "space star*" gives the following error:
----
shell-command: Setting current directory: No such file or directory, /home/tdw/space\ star\*/
----

arduino-cli-board-list now gives the following message:
----
Port         Protocol Type              Board Name FQBN Core
/dev/ttyUSB0 serial   Serial Port (USB) Unknown

----
(note blank lines at the bottom).  All the commands that use arduino-cli--message show the blank lines at the bottom, so I string-trim the message first to eliminate them.

tdwilson committed:
Remove redundant expand-file-name from arduino-cli--message

The expand-file-name in the previous commit is not required
as long as the PATH (if provided) is an absolute directory name.

tdwilson committed:
Merge branch 'master' into default-directory-may-start-with-tilde

ahihi commented:
continuing the discussion from #19 (btw for future reference, its not necessary to create a new PR - you can push new commits to an existing one, or even rewrite the PR history by force-pushing if needed :)

> I think your example still uses an absolute directory name, but I was thinking of the user running `arduino-cli-new-sketch` and entering "../" as the directory file name (which is reasonable, indicating that the user wants to create the new sketch in a directory parallel to the current sketch).

ah, i see. then yes, it looks like we do need `expand-file-name` in `arduino-cli--message`.

> Emacs doesn't object to `arduino-cli--message` setting `default-directory` to this (illegal) relative directory file name, but the result of the `shell-command-to-string` is `Error creating sketch: Can’t create sketch: mkdir /test: permission denied`. (I haven't looked into why, but as I put an illegal value into `default-directory`, I guess Emacs is allowed to do what it wants when it uses the value.)

in this case, i think the `..` is resolved relative to the Emacs process' own cwd, which will depend on how it was launched and indeed is not very intuitive behavior.

> When I put `expand-file-name` into `arduino-cli-new-sketch` and again enter "../" as the directory for the new sketch, it all works as I expect. (I'll commit my version when I have time over the next few days - it also makes the parent directory the default result for `read-directory-name`, as putting the new sketch in a parallel directory to the current sketch seems the most obvious use case to me.)

i think this last part is rather opinionated and i would not make it the default; it deviates from conventional Emacs behavior which is to start file/directory selection in `default-directory`. instead, you might implement it in your own config via advice, e.g.:

```elisp
(defun my-arduino-cli-new-sketch-advice (orig-fun &rest args)
  (let ((default-directory (file-name-parent-directory default-directory)))
    (apply orig-fun args)))

(advice-add #'arduino-cli-new-sketch :around #'my-arduino-cli-new-sketch-advice)
```

tdwilson committed:
Merge branch 'master' into default-directory-may-start-with-tilde

tdwilson committed:
Ensure PATH for arduino-cli--message is an absolute directory name

arduino-cli--message requires an absolute directory name as its second
argument, but the result of read-directory-name will be a relative
path name, such as `../`, if that is what the user enters.

tdwilson commented:
Thank you for the `advice-add tip`, as I do prefer that behaviour: I can understand that experienced Emacs users may prefer for `arduino-cli-new-sketch` to offer the "default default directory" in accordance with the principle of least surprise. For me, coming from Arduino IDE, the behaviour similar to Arduino IDE is more intuitive, and for me the surprise is to be offered a default that is unlikely to be what I want -- rather, without your advice-add tip, I would have to remember to back up a directory level each time.

I think I've finished this branch now, as `arduino-cli-new-sketch`, `arduino-cli-compile` and the `arduino-cli-`*`-list` commands are working for me on Linux.  I haven't yet tried all the other arduino-cli commands, but I'd be happy to create new branches if I have tweaks to propose when I try them.

ahihi commented:
thank you, merging!

the problem with making the parent directory the default for everyone is that it assumes `arduino-cli-new-sketch` will be run in the context of an existing Arduino sketch, within a directory that other sketches should go into. but this may not be the case - for example, i often run it from Dired or some non-Arduino source file within the directory tree of a larger project, to introduce an Arduino sketch into the project.

there might be a reasonable argument that we should have a customization variable for the default sketch directory, like "Sketchbook location" in the Arduino IDE settings. but unless more people request this, i think for now this is a good use case for advice :)
